contrib/repro_test.bzl

```diff
@@ -0,0 +1,243 @@
+
+    # Build the Bazel image to build and repro the test_image in.
+    # Mount the source project and its path into this image.
+    _container.image.implementation(
+        ctx,
+        files = [proj_root],
+        tars = [proj_tar],
+        output_executable = ctx.actions.declare_file(name + "_load.sh"),
+        output_tarball = image_tar,
+        workdir = "/",
+    )
+
+    test_img_label = ctx.attr.test_image.label
+    build_targets = [
+        str(test_img_label),
+        str(test_img_label) + ".tar",
+        str(test_img_label) + ".digest",
+        str(test_img_label) + ".json",
+    ]
+    extract_path = "/img_outs"
+    commands = [
+        "cd \$(cat /%s)" % proj_root.basename,
+        "bazel build " + " ".join(build_targets),
+        "cp -r bazel-bin/%s %s" % (test_img_label.package, extract_path),
+    ]
+    docker_run_flags = ["--entrypoint", "''"]
+
+    img_outs = ctx.outputs.img_outs
+    img_outs_repro = ctx.outputs.img_outs_repro
+
+    # Build and extract test_image inside a container.
+    _extract.implementation(
+        ctx,
+        name = "build_test_image_and_extract",
+        commands = commands,
+        extract_file = extract_path,
+        image = image_tar,
+        docker_run_flags = docker_run_flags,
+        output_file = img_outs,
+        script_file = ctx.actions.declare_file(name + ".build"),
+    )
+
+    # Build and extract test_image again in a different container to try and
+    # repro previously built test_image.
+    _extract.implementation(
+        ctx,
+        name = "build_test_image_and_extract_repro",
+        commands = commands,
+        extract_file = extract_path,
+        image = image_tar,
+        docker_run_flags = docker_run_flags,
+        output_file = img_outs_repro,
+        script_file = ctx.actions.declare_file(name + "_repro.build"),
+    )
+
+    # Expand template to run the image comparison test.
+    diff_tool_exec = ctx.executable._container_diff_tool
+    ctx.actions.expand_template(
+        template = ctx.file._test_tpl,
+        substitutions = {
+            "%{container_diff_args}": " ".join(ctx.attr.container_diff_args),
+            "%{container_diff_tool}": diff_tool_exec.short_path,
+            "%{img1_path}": img_outs.short_path,
+            "%{img2_path}": img_outs_repro.short_path,
+            "%{img_name}": test_img_label.name,
+        },
+        output = ctx.outputs.test_script,
+        is_executable = True,
+    )
+
+    runfiles = ctx.runfiles(files = [img_outs, img_outs_repro, diff_tool_exec])
+
+    return [
+        DefaultInfo(
+            executable = ctx.outputs.test_script,
+            runfiles = runfiles,
+        ),
+    ]
+
+container_repro_test = rule(
+    attrs = dicts.add(_container.image.attrs, {
+        "test_image": attr.label(
+            doc = "A container_image target to test for reproducibility. " +
+                  "Must be a label relative to the root of the project and " +
```

Comment:
use https://docs.bazel.build/versions/master/skylark/lib/attr.html#label allow_rules to list this can only receive container_image and toolchain_container targets).
You should also be able to lift this restriction (see above where ctx.attr.test_image is used)

Reply:
Done